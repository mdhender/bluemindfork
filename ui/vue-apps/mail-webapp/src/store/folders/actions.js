import { ItemFlag } from "@bluemind/core.container.api";
import api from "../api/apiFolders";
import {
    ADD_FOLDERS,
    ADD_FOLDER,
    //TODO: change mutation names
    REMOVE_FOLDER as MUTATION_REMOVE_FOLDER,
    RENAME_FOLDER as MUTATION_RENAME_FOLDER
} from "./mutations";
import { FolderAdaptor } from "./helpers/FolderAdaptor";

export const FETCH_FOLDERS = "FETCH_FOLDERS";
const fetchFolders = async function ({ commit }, mailbox) {
    const items = await api.getAllFolders(mailbox);
    const folders = items
        .filter(item => !item.flags.includes(ItemFlag.Deleted))
        .sort((a, b) => a.value.fullName.toLowerCase().localeCompare(b.value.fullName.toLowerCase()))
        .map(item => FolderAdaptor.fromMailboxFolder(item, mailbox));
    commit(ADD_FOLDERS, folders);
};

export const CREATE_FOLDER = "CREATE_FOLDER";
const createFolder = async function ({ commit, state }, { key, name, parent, mailbox }) {
    const foldertoadd = FolderAdaptor.create(key, name, parent && state[parent], mailbox);
    commit(ADD_FOLDER, foldertoadd);
    const item = FolderAdaptor.toMailboxFolder(foldertoadd, mailbox);
    try {
        const { uid, id } = await api.createNewFolder(mailbox, item);
        commit(ADD_FOLDER, { ...foldertoadd, uid, id });
    } catch (e) {
        commit(MUTATION_REMOVE_FOLDER, foldertoadd.key);
        throw e;
    }
};

export const REMOVE_FOLDER = "REMOVE_FOLDER";
const removeFolder = async function ({ state, commit }, { key, mailbox }) {
    const folder = state[key];
    commit(MUTATION_REMOVE_FOLDER, folder.key);
    try {
        await api.deleteFolder(mailbox, folder);
    } catch (e) {
        commit(ADD_FOLDER, folder);
        throw e;
    }
};

export const RENAME_FOLDER = "RENAME_FOLDER";
const renameFolder = async function ({ commit, state }, { folder, mailbox }) {
    const { name: oldName, path: oldPath } = state[folder.key];
    commit(MUTATION_RENAME_FOLDER, { name: folder.name, key: folder.key, path: folder.path });
    const item = FolderAdaptor.toMailboxFolder(folder, mailbox);
    try {
        await api.updateFolder(mailbox, item);
    } catch (e) {
        commit(MUTATION_RENAME_FOLDER, { name: oldName, key: folder.key, path: oldPath });
        throw e;
    }
};
export default {
    [FETCH_FOLDERS]: fetchFolders,
    [CREATE_FOLDER]: createFolder,
    // FIXME: when deleting a folder having children, it is not deleted in UI & we got errors in the console..
    [REMOVE_FOLDER]: removeFolder,
    [RENAME_FOLDER]: renameFolder
};
