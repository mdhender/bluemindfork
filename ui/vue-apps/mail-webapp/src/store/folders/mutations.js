import Vue from "vue";

export const ADD_FOLDER = "ADD_FOLDER";
const addFolder = (state, { key, ...folder }) => {
    Vue.set(state, key, { ...folder, key });
};

export const ADD_FOLDERS = "ADD_FOLDERS";
const addFolders = (state, folders) => {
    folders.forEach(folder => {
        addFolder(state, folder);
    });
};

export const RENAME_FOLDER = "RENAME_FOLDER";
const renameFolder = (state, { key, name, path }) => {
    state[key].name = name;
    state[key].path = path;
};

//TODO: change signature to: ({key}) => {}
export const REMOVE_FOLDER = "REMOVE_FOLDER";
const removeFolder = function (state, key) {
    Vue.delete(state, key);
};

//TODO: change parameter name s/count/unread/
export const SET_UNREAD_COUNT = "SET_UNREAD_COUNT";
const setUnreadCount = (state, { key, count }) => {
    // FIXME: [inconsistency] With `ADD_FOLDER` we can set a negative value for the `unread` property
    if (count >= 0 && state[key].unread !== count) {
        state[key].unread = count;
    }
};

//TODO: change signature to: ({key}) => {}
export const TOGGLE_FOLDER = "TOGGLE_FOLDER";
const toggleFolder = (state, key) => {
    state[key].expanded = !state[key].expanded;
};

export default {
    [ADD_FOLDER]: addFolder,
    [ADD_FOLDERS]: addFolders,
    [RENAME_FOLDER]: renameFolder,
    [REMOVE_FOLDER]: removeFolder,
    [SET_UNREAD_COUNT]: setUnreadCount,
    [TOGGLE_FOLDER]: toggleFolder
};
