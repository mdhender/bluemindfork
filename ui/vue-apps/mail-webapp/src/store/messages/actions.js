import api from "../api/apiMessages";
import {
    ADD_MESSAGES,
    MARK_AS_READ as MUTATION_MARK_AS_READ,
    MARK_AS_UNREAD as MUTATION_MARK_AS_UNREAD
} from "./mutations";

export const FETCH_MESSAGES = "FETCH_MESSAGES";
const fetchMessages = async ({ commit }, { folderUid }) => {
    try {
        const messages = await api.fetchMessages({ folderUid });
        commit(
            ADD_MESSAGES,
            messages.map(message => ({
                ...message,
                key: message.uid,
                folder: folderUid
            }))
        );
    } catch (error) {
        console.error("unable to fetch messages", error);
    }
};

export const MARK_AS_READ = "MARK_AS_READ";
const markAsRead = async ({ commit }, { key, folder }) => {
    commit(MUTATION_MARK_AS_READ, { key });
    try {
        await api.markAsRead({ uid: folder }, { key });
    } catch (error) {
        commit(MUTATION_MARK_AS_UNREAD, { key });
    }
};

export default {
    [FETCH_MESSAGES]: fetchMessages,
    [MARK_AS_READ]: markAsRead
};
