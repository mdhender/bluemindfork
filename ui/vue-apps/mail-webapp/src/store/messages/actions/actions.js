import mutationsTypes from "../../mutationTypes";
import { MessageStatus, partialCopy } from "../../../model/message";
import apiMessages from "../../api/apiMessages";

export async function addFlag({ commit, getters, state }, { messageKeys, flag }) {
    const keys = Array.isArray(messageKeys) ? messageKeys : [messageKeys];
    const messages = keys
        .map(key => state[key])
        .filter(message => !getters.isLoaded(message.key) || !message.flags.includes(flag));
    commit(mutationsTypes.ADD_FLAG, { keys, flag });
    try {
        await apiMessages.addFlag(messages, flag);
    } catch (e) {
        commit(mutationsTypes.DELETE_FLAG, { keys, flag });
        throw e;
    }
}

export async function deleteFlag({ commit, state, getters }, { messageKeys, flag }) {
    const keys = Array.isArray(messageKeys) ? messageKeys : [messageKeys];
    const messages = keys
        .map(key => state[key])
        .filter(message => !getters.isLoaded(message.key) || message.flags.includes(flag));
    commit(mutationsTypes.DELETE_FLAG, { keys: messages.map(({ key }) => key), flag });
    try {
        await apiMessages.deleteFlag(messages, flag);
    } catch (e) {
        commit(mutationsTypes.ADD_FLAG, { keys: messages.map(({ key }) => key), flag });
        throw e;
    }
}

export async function fetchMessageMetadata({ commit, state }, { messageKeys }) {
    const messages = (Array.isArray(messageKeys) ? messageKeys : [messageKeys]).map(key => state[key]);
    const fullMessages = await apiMessages.multipleById(messages);
    commit(mutationsTypes.ADD_MESSAGES, fullMessages);
}

export async function removeMessages({ commit, state }, messageKeys) {
    const messages = (Array.isArray(messageKeys) ? messageKeys : [messageKeys]).map(key => partialCopy(state[key]));
    commit(
        mutationsTypes.SET_MESSAGES_STATUS,
        messages.map(({ key }) => ({ key, status: MessageStatus.REMOVED }))
    );

    try {
        await apiMessages.multipleDeleteById(messages);
        commit(
            mutationsTypes.REMOVE_MESSAGES,
            messages.map(({ key }) => key)
        );
    } catch (e) {
        commit(mutationsTypes.SET_MESSAGES_STATUS, messages);
        throw e;
    }
}
