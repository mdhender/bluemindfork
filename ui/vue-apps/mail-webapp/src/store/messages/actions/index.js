import { withAlert } from "../../helpers/withAlert";

import {
    addFlag,
    deleteFlag,
    fetchMessageIfNotLoaded,
    fetchMessageMetadata,
    moveMessages,
    removeMessages
} from "./actions";
import addAttachments from "./addAttachments";
import removeAttachment from "./removeAttachment";
import { debouncedSave, saveAsap } from "./save";
import send from "./send";
import {
    ADD_ATTACHMENTS,
    ADD_FLAG,
    DEBOUNCED_SAVE_MESSAGE,
    DELETE_FLAG,
    FETCH_MESSAGE_IF_NOT_LOADED,
    FETCH_MESSAGE_METADATA,
    MARK_MESSAGE_AS_FLAGGED,
    MARK_MESSAGE_AS_READ,
    MARK_MESSAGE_AS_UNFLAGGED,
    MARK_MESSAGE_AS_UNREAD,
    MARK_MESSAGES_AS_FLAGGED,
    MARK_MESSAGES_AS_READ,
    MARK_MESSAGES_AS_UNFLAGGED,
    MARK_MESSAGES_AS_UNREAD,
    MOVE_MESSAGES,
    REMOVE_ATTACHMENT,
    REMOVE_MESSAGES,
    SAVE_MESSAGE,
    SAVE_AS_DRAFT,
    SAVE_AS_TEMPLATE,
    SEND_MESSAGE
} from "~/actions";

import { Flag } from "@bluemind/email";

const markAsUnread = ({ dispatch }, messages) => dispatch(DELETE_FLAG, { messages, flag: Flag.SEEN });
const markAsRead = ({ dispatch }, messages) => dispatch(ADD_FLAG, { messages, flag: Flag.SEEN });
const markAsFlagged = ({ dispatch }, messages) => dispatch(ADD_FLAG, { messages, flag: Flag.FLAGGED });
const markAsUnflagged = ({ dispatch }, messages) => dispatch(DELETE_FLAG, { messages, flag: Flag.FLAGGED });
const saveAs = (context, { message, messageCompose }) => saveAsap(context, { draft: message, messageCompose });
export default {
    [ADD_ATTACHMENTS]: addAttachments,
    [ADD_FLAG]: addFlag,
    [DEBOUNCED_SAVE_MESSAGE]: debouncedSave,
    [DELETE_FLAG]: deleteFlag,
    [FETCH_MESSAGE_IF_NOT_LOADED]: fetchMessageIfNotLoaded,
    [FETCH_MESSAGE_METADATA]: fetchMessageMetadata,
    [MARK_MESSAGE_AS_FLAGGED]: markAsFlagged,
    [MARK_MESSAGE_AS_READ]: markAsRead,
    [MARK_MESSAGE_AS_UNFLAGGED]: markAsUnflagged,
    [MARK_MESSAGE_AS_UNREAD]: markAsUnread,
    [MARK_MESSAGES_AS_FLAGGED]: withAlert(markAsFlagged, MARK_MESSAGES_AS_FLAGGED),
    [MARK_MESSAGES_AS_READ]: withAlert(markAsRead, MARK_MESSAGES_AS_READ),
    [MARK_MESSAGES_AS_UNFLAGGED]: withAlert(markAsUnflagged, MARK_MESSAGES_AS_UNFLAGGED),
    [MARK_MESSAGES_AS_UNREAD]: withAlert(markAsUnread, MARK_MESSAGES_AS_UNREAD),
    [MOVE_MESSAGES]: withAlert(moveMessages, MOVE_MESSAGES, "MoveMessages"),
    [REMOVE_ATTACHMENT]: removeAttachment,
    [REMOVE_MESSAGES]: withAlert(removeMessages, REMOVE_MESSAGES, "RemoveMessages"),
    [SAVE_MESSAGE]: saveAsap,
    [SAVE_AS_TEMPLATE]: withAlert(saveAs, SAVE_AS_TEMPLATE, "SaveMessageAs"),
    [SAVE_AS_DRAFT]: withAlert(saveAs, SAVE_AS_DRAFT, "SaveMessageAs"),
    [SEND_MESSAGE]: withAlert(send, SEND_MESSAGE, "SendMessage")
};
