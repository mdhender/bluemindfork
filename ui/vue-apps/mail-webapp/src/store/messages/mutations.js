import Vue from "vue";

export const ADD_MESSAGE = "ADD_MESSAGE";
const addMessage = (state, { key, ...message }) => {
    Vue.set(state, key, { ...message, key });
};

export const ADD_MESSAGES = "ADD_MESSAGES";
const addMessages = (state, messages) => {
    messages.forEach(message => {
        addMessage(state, message);
    });
};

export const UPDATE_STATUS = "UPDATE_STATUS";
const updateStatus = (state, { key, status }) => {
    const message = state[key];
    if (message) {
        message.status = status;
    }
};

export const MARK_AS_READ = "MARK_AS_READ";
const markAsRead = updateReadFlag(true);

export const MARK_AS_UNREAD = "MARK_AS_UNREAD";
const markAsUnread = updateReadFlag(false);

function updateReadFlag(isRead) {
    return (state, { key }) => {
        const message = state[key];
        if (message) {
            const flags = message.data && message.data.flags;
            const newLocal = {
                ...message.data,
                flags: { ...flags, read: isRead }
            };
            message.data = newLocal;
        }
    };
}

export default {
    [ADD_MESSAGE]: addMessage,
    [ADD_MESSAGES]: addMessages,
    [UPDATE_STATUS]: updateStatus,
    [MARK_AS_READ]: markAsRead,
    [MARK_AS_UNREAD]: markAsUnread
};

//TODO: Voici les mutations qui sont utilisées ci et là
// const addFlag = (state, { messageKeys, mailboxItemFlag }) => {};
// const deleteFlag = (state, { messageKeys, mailboxItemFlag }) => {};
// const storePartContent = (state, { messageKey, address, content }) => {};
// const storeItems = (state, { items, folderUid }) => {};
// const setItemKeysByIdsFolderUid = (state, { ids, folderUid }) => {};
// const setItemKeys = (state, itemKeys) => {};
// const removeItems = (state, messageKeys) => {};
