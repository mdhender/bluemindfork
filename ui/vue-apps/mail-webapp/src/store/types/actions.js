// partsData
export const FETCH_PART_DATA = "FETCH_PART_DATA";

// mailboxes
export const FETCH_MAILBOXES = "FETCH_MAILBOXES";

// folders
export const CREATE_FOLDER = "CREATE_FOLDER";
export const CREATE_FOLDER_HIERARCHY = "CREATE_FOLDER_HIERARCHY";
export const EMPTY_FOLDER = "EMPTY_FOLDER";
export const EMPTY_TRASH = "EMPTY_TRASH";
export const FETCH_FOLDERS = "FETCH_FOLDERS";
export const MARK_FOLDER_AS_READ = "MARK_FOLDER_AS_READ";
export const MOVE_FOLDER = "MOVE_FOLDER";
export const REMOVE_FOLDER = "REMOVE_FOLDER";
export const RENAME_FOLDER = "RENAME_FOLDER";
export const UNREAD_FOLDER_COUNT = "$_VueBus_mail-webapp_unread_folder_count";
export const PUSHED_FOLDER_CHANGES = "$_VueBus_mail-webapp_pushed_folder_changes";

// messages
export const ADD_ATTACHMENT = "ADD_ATTACHMENT";
export const ADD_FLAG = "ADD_FLAG";
export const ADD_LOCAL_ATTACHMENT = "ADD_LOCAL_ATTACHMENT";
export const DEBOUNCED_SAVE_MESSAGE = "DEBOUNCED_SAVE_MESSAGE";
export const DEBOUNCED_SET_MESSAGE_CONTENT = "DEBOUNCED_SET_MESSAGE_CONTENT";
export const DELETE_FLAG = "DELETE_FLAG";
export const FETCH_MESSAGE_IF_NOT_LOADED = "FETCH_MESSAGE_IF_NOT_LOADED";
export const FETCH_MESSAGE_METADATA = "FETCH_MESSAGE_METADATA";
export const IMPORT_EML = "IMPORT_EML";
export const MARK_MESSAGE_AS_FLAGGED = "MARK_MESSAGE_AS_FLAGGED";
export const MARK_MESSAGE_AS_READ = "MARK_MESSAGE_AS_READ";
export const MARK_MESSAGE_AS_UNFLAGGED = "MARK_MESSAGE_AS_UNFLAGGED";
export const MARK_MESSAGE_AS_UNREAD = "MARK_MESSAGE_AS_UNREAD";
export const MARK_MESSAGES_AS_FLAGGED = "MARK_MESSAGES_AS_FLAGGED";
export const UNEXPUNGE = "UNEXPUNGE";
export const MARK_MESSAGES_AS_READ = "MARK_MESSAGES_AS_READ";
export const MARK_MESSAGES_AS_UNFLAGGED = "MARK_MESSAGES_AS_UNFLAGGED";
export const MARK_MESSAGES_AS_UNREAD = "MARK_MESSAGES_AS_UNREAD";
export const MOVE_MESSAGES = "MOVE_MESSAGES";
export const REMOVE_ATTACHMENT = "REMOVE_ATTACHMENT";
export const REMOVE_MESSAGES = "REMOVE_MESSAGES";
export const REQUEST_DSN = "REQUEST_DSN";
export const SAVE_AS_DRAFT = "SAVE_AS_DRAFT";
export const SAVE_AS_TEMPLATE = "SAVE_AS_TEMPLATE";
export const SAVE_MESSAGE = "SAVE_MESSAGE";
export const SEND_MESSAGE = "SEND_MESSAGE";
export const TOGGLE_DSN_REQUEST = "TOGGLE_DSN_REQUEST";
export const UPDATE_MESSAGE_STRUCTURE = "UPDATE_MESSAGE_STRUCTURE";

// conversationList
export const FETCH_CONVERSATION_LIST_KEYS = "FETCH_CONVERSATION_LIST_KEYS";
export const CONVERSATION_LIST_NEXT_PAGE = "CONVERSATION_LIST_NEXT_PAGE";
export const REFRESH_CONVERSATION_LIST_KEYS = "REFRESH_CONVERSATION_LIST_KEYS";

// messageCompose
export const FETCH_TEMPLATES_KEYS = "FETCH_TEMPLATES_KEYS";
export const LOAD_MAX_MESSAGE_SIZE = "LOAD_MAX_MESSAGE_SIZE";
export const SET_DRAFT_CONTENT = "SET_DRAFT_CONTENT";
export const SET_MESSAGE_CONTENT = "SET_MESSAGE_CONTENT";

// consultPanel
export const FETCH_EVENT = "FETCH_EVENT";
export const SET_EVENT_STATUS = "SET_EVENT_STATUS";
export const ACCEPT_COUNTER_EVENT = "ACCEPT_COUNTER_EVENT";
export const DECLINE_COUNTER_EVENT = "DECLINE_COUNTER_EVENT";
export const REJECT_ATTENDEES = "REJECT_ATTENDEES";
// conversations
export const FETCH_CONVERSATION_IF_NOT_LOADED = "FETCH_CONVERSATION_IF_NOT_LOADED";
export const FETCH_CONVERSATIONS = "FETCH_CONVERSATIONS";
export const MARK_CONVERSATIONS_AS_FLAGGED = "MARK_CONVERSATIONS_AS_FLAGGED";
export const MARK_CONVERSATIONS_AS_READ = "MARK_CONVERSATIONS_AS_READ";
export const MARK_CONVERSATIONS_AS_UNFLAGGED = "MARK_CONVERSATIONS_AS_UNFLAGGED";
export const MARK_CONVERSATIONS_AS_UNREAD = "MARK_CONVERSATIONS_AS_UNREAD";
export const MOVE_CONVERSATION_MESSAGES = "MOVE_CONVERSATION_MESSAGES";
export const MOVE_CONVERSATIONS = "MOVE_CONVERSATIONS";
export const REMOVE_CONVERSATION_MESSAGES = "REMOVE_CONVERSATION_MESSAGES";
export const REMOVE_CONVERSATIONS = "REMOVE_CONVERSATIONS";

// folderList
export const RESET_FILTER = "RESET_FILTER";
export const SHOW_MORE_FOR_GROUP_MAILBOXES = "SHOW_MORE_FOR_GROUP_MAILBOXES";
export const SHOW_MORE_FOR_MAILSHARES = "SHOW_MORE_FOR_MAILSHARES";
export const SHOW_MORE_FOR_USERS = "SHOW_MORE_FOR_USERS";

// search
export const RESET_CURRENT_SEARCH = "RESET_CURRENT_SEARCH";

//preview
export const SET_PREVIEW = "SET_PREVIEW";
