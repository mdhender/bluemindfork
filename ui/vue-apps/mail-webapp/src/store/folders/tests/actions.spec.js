import Vue from "vue";
import Vuex from "vuex";
import deepClone from "lodash.clonedeep";
import api from "../../api/apiFolders";
import storeConfig from "../index";
import { FETCH_FOLDERS, CREATE_FOLDER, REMOVE_FOLDER, RENAME_FOLDER } from "../actions";
import { ADD_FOLDER } from "../mutations";

Vue.use(Vuex);
jest.mock("../../api/apiFolders");

describe("actions", () => {
    test("Fetch folders", async () => {
        const store = new Vuex.Store(deepClone(storeConfig));
        const mailbox = {
            type: "foo",
            name: "bar"
        };
        api.getAllFolders.mockReturnValue(require("../../tests/data/users/alice/folders.json"));
        await store.dispatch(FETCH_FOLDERS, mailbox);
        expect(store.state).toMatchSnapshot();
    });
    test("Create fetched folder", async () => {
        const store = new Vuex.Store(deepClone(storeConfig));
        const mailbox = {
            type: "",
            name: ""
        };
        const folder = { key: "foo", name: "bar", path: "baz" };
        api.createNewFolder.mockImplementation((mailbox, item) => {
            if (
                item.uid === folder.key &&
                item.value.name === folder.name &&
                item.value.fullName === folder.name &&
                item.value.parentUid === null
            ) {
                return Promise.resolve({ uid: `${item.value.name}-baz`, id: `${item.value.name}-baz` });
            } else {
                return Promise.reject("Invalid parameters");
            }
        });
        await store.dispatch(CREATE_FOLDER, { ...folder, mailbox });
        expect(Object.keys(store.state)).toHaveLength(1);
        expect(store.state).toMatchSnapshot();
    });
    test("Create a folder with failure", async () => {
        const store = new Vuex.Store(deepClone(storeConfig));
        const mailbox = {
            type: "",
            name: ""
        };
        api.createNewFolder.mockRejectedValue(new Error("Mocked rejection"));
        try {
            await store.dispatch(CREATE_FOLDER, {
                ...{ key: "foo", name: "Foo", path: "foo", parent: null },
                mailbox
            });
        } catch (error) {
            expect(error.message).toEqual("Mocked rejection");
            expect(store.state).toEqual({});
        }
    });
    test("Rename folder", async () => {
        const store = new Vuex.Store(deepClone(storeConfig));
        const mailbox = {
            type: "",
            name: ""
        };
        const oldFolder = { key: "1", name: "foo", path: "foobaz" };
        const newFolder = { key: "1", name: "bar", path: "barbaz" };
        api.updateFolder.mockImplementation((mailbox, item) => {
            if (
                item.value.name === newFolder.name &&
                item.value.fullName === newFolder.path &&
                item.value.parentUid === undefined
            )
                return Promise.resolve();
            else return Promise.reject("Invalid parameters");
        });
        store.commit(ADD_FOLDER, oldFolder);
        await store.dispatch(RENAME_FOLDER, { folder: newFolder, mailbox });
        expect(store.state).toEqual({ [newFolder.key]: newFolder });
    });
    test("Rename folder with failure", async () => {
        const store = new Vuex.Store(deepClone(storeConfig));
        const mailbox = { type: "", name: "" };
        const oldFolder = { key: "1", name: "foo", path: "baz" };
        const newFolder = { key: "1", name: "bar", path: "foobar" };
        api.updateFolder.mockRejectedValue(new Error("Mocked rejection"));
        store.commit(ADD_FOLDER, oldFolder);
        try {
            await store.dispatch(RENAME_FOLDER, { folder: newFolder, mailbox });
        } catch (error) {
            expect(error.message).toEqual("Mocked rejection");
            expect(store.state).toEqual({ [oldFolder.key]: oldFolder });
        }
    });
    test("Remove folder with optimistic return", async () => {
        const store = new Vuex.Store(deepClone(storeConfig));
        const mailbox = { type: "", name: "" };
        const folder1 = { key: "1", name: "foo", path: "baz" };
        store.commit(ADD_FOLDER, folder1);
        const folder2 = { key: "2", name: "bar", path: "baz" };
        store.commit(ADD_FOLDER, folder2);
        api.deleteFolder.mockImplementation((mailbox, folder) => {
            if (folder.key === folder1.key) {
                expect(store.state).toEqual({ [folder2.key]: folder2 });
                return {};
            }
            throw new Error("Test Error: invalid parameters");
        });
        await store.dispatch(REMOVE_FOLDER, { key: folder1.key, mailbox });
        expect(store.state).toEqual({ [folder2.key]: folder2 });
    });
    test("Remove folder optimistic return with failure", async () => {
        const store = new Vuex.Store(deepClone(storeConfig));
        const mailbox = { type: "", name: "" };
        const folder1 = { key: "1", name: "foo", path: "baz" };
        store.commit(ADD_FOLDER, folder1);
        const folder2 = { key: "2", name: "bar", path: "baz" };
        store.commit(ADD_FOLDER, folder2);
        api.deleteFolder.mockImplementation((mailbox, folder) => {
            if (folder.key === folder1.key) {
                expect(store.state).toEqual({ [folder2.key]: folder2 });
                return {};
            }
            throw new Error("Test Error: invalid parameters");
        });
        try {
            await store.dispatch(REMOVE_FOLDER, { key: folder1.key, mailbox });
        } catch (error) {
            expect(error.message).toEqual("Mocked rejection");
            expect(store.state).toEqual({ [folder1.key]: folder1, [folder2.key]: folder2 });
        }
    });
});
