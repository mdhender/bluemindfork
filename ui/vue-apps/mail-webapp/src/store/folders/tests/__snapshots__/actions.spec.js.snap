// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions Create fetched folder 1`] = `
Object {
  "foo": Object {
    "default": true,
    "expanded": false,
    "id": "bar-baz",
    "imapName": "bar",
    "key": "foo",
    "mailbox": undefined,
    "name": "bar",
    "parent": null,
    "path": "bar",
    "uid": "bar-baz",
    "unread": 0,
    "writable": undefined,
  },
}
`;

exports[`actions Fetch folders 1`] = `
Object {
  "050ca560-37ae-458a-bd52-c40fedf4068d": Object {
    "default": true,
    "expanded": false,
    "id": 462,
    "imapName": "2017",
    "key": "050ca560-37ae-458a-bd52-c40fedf4068d",
    "mailbox": undefined,
    "name": "2017",
    "parent": "135adc10-db84-440e-aebc-e10d185fa227",
    "path": "Archives/2017",
    "uid": "050ca560-37ae-458a-bd52-c40fedf4068d",
    "unread": 0,
    "writable": undefined,
  },
  "135adc10-db84-440e-aebc-e10d185fa227": Object {
    "default": true,
    "expanded": false,
    "id": 460,
    "imapName": "Archives",
    "key": "135adc10-db84-440e-aebc-e10d185fa227",
    "mailbox": undefined,
    "name": "Archives",
    "parent": null,
    "path": "Archives",
    "uid": "135adc10-db84-440e-aebc-e10d185fa227",
    "unread": 0,
    "writable": undefined,
  },
  "65b51768-8544-4bb3-9950-c6f4fc32a52e": Object {
    "default": true,
    "expanded": false,
    "id": 464,
    "imapName": "Clients",
    "key": "65b51768-8544-4bb3-9950-c6f4fc32a52e",
    "mailbox": undefined,
    "name": "Clients",
    "parent": "050ca560-37ae-458a-bd52-c40fedf4068d",
    "path": "Archives/2017/Clients",
    "uid": "65b51768-8544-4bb3-9950-c6f4fc32a52e",
    "unread": 0,
    "writable": undefined,
  },
  "67201a3a-91cb-4198-a153-85a8174be9e6": Object {
    "default": true,
    "expanded": false,
    "id": 427,
    "imapName": "Drafts",
    "key": "67201a3a-91cb-4198-a153-85a8174be9e6",
    "mailbox": undefined,
    "name": "Drafts",
    "parent": null,
    "path": "Drafts",
    "uid": "67201a3a-91cb-4198-a153-85a8174be9e6",
    "unread": 0,
    "writable": undefined,
  },
  "7998fa63-3854-4ecb-8558-9596065c0128": Object {
    "default": true,
    "expanded": false,
    "id": 423,
    "imapName": "Sent",
    "key": "7998fa63-3854-4ecb-8558-9596065c0128",
    "mailbox": undefined,
    "name": "Sent",
    "parent": null,
    "path": "Sent",
    "uid": "7998fa63-3854-4ecb-8558-9596065c0128",
    "unread": 0,
    "writable": undefined,
  },
  "98a9383e-5156-44eb-936a-e6b0825b0809": Object {
    "default": true,
    "expanded": false,
    "id": 425,
    "imapName": "Trash",
    "key": "98a9383e-5156-44eb-936a-e6b0825b0809",
    "mailbox": undefined,
    "name": "Trash",
    "parent": null,
    "path": "Trash",
    "uid": "98a9383e-5156-44eb-936a-e6b0825b0809",
    "unread": 0,
    "writable": undefined,
  },
  "eb4cc101-60bf-48ac-91ce-7698bd29ba73": Object {
    "default": true,
    "expanded": false,
    "id": 431,
    "imapName": "Outbox",
    "key": "eb4cc101-60bf-48ac-91ce-7698bd29ba73",
    "mailbox": undefined,
    "name": "Outbox",
    "parent": null,
    "path": "Outbox",
    "uid": "eb4cc101-60bf-48ac-91ce-7698bd29ba73",
    "unread": 0,
    "writable": undefined,
  },
  "f1c3f42f-551b-446d-9682-cfe0574b3205": Object {
    "default": true,
    "expanded": false,
    "id": 421,
    "imapName": "INBOX",
    "key": "f1c3f42f-551b-446d-9682-cfe0574b3205",
    "mailbox": undefined,
    "name": "INBOX",
    "parent": null,
    "path": "INBOX",
    "uid": "f1c3f42f-551b-446d-9682-cfe0574b3205",
    "unread": 0,
    "writable": undefined,
  },
  "fd965e68-c276-4db0-a298-01a92d3c9822": Object {
    "default": true,
    "expanded": false,
    "id": 429,
    "imapName": "Junk",
    "key": "fd965e68-c276-4db0-a298-01a92d3c9822",
    "mailbox": undefined,
    "name": "Junk",
    "parent": null,
    "path": "Junk",
    "uid": "fd965e68-c276-4db0-a298-01a92d3c9822",
    "unread": 0,
    "writable": undefined,
  },
}
`;
